packages/bundle-source/src/fs.js

```diff
@@ -0,0 +1,81 @@
+/**
+ * @param {string} fileName
+ * @param {{
+ *   fs: {
+ *     promises: Pick<import('fs/promises'),'readFile' | 'stat'>
+ *   },
+ *   path: Pick<import('path'), 'resolve' | 'relative' | 'normalize'>,
+ * }} powers
+ */
+export const makeFileReader = (fileName, { fs, path }) => {
+  const make = there => makeFileReader(there, { fs, path });
+
+  // fs.promises.exists isn't implemented in Node.js apparently because it's pure
+  // sugar.
+  const exists = fn =>
+    fs.promises.stat(fn).then(
+      () => true,
+      e => {
+        if (e.code === 'ENOENT') {
+          return false;
+        }
+        throw e;
+      },
+    );
+
+  const readText = () => fs.promises.readFile(fileName, 'utf-8');
+
+  const maybeReadText = () =>
+    readText().catch(error => {
+      if (
+        error.message.startsWith('ENOENT: ') ||
+        error.message.startsWith('EISDIR: ')
+      ) {
+        return undefined;
+      }
+      throw error;
+    });
```

Comment:
Using `maybeReadText` instead of `exists` eliminates the race to read the `meta` file after checking that it exists. This was not a problem under the lock, but becomes a problem when we rely on atomic writes instead.